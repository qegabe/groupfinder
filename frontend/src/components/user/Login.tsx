import React from "react";
import { Alert, Box, Button, TextField, Typography } from "@mui/material";
import useFormData from "../../hooks/useFormData";
import { useAppDispatch, useAppSelector } from "../../hooks/redux";
import { login } from "../../actions/actionCreators";
import parseFormErrors from "../../helpers/parseFormErrors";

function Login() {
  const [formData, handleChange] = useFormData({ username: "", password: "" });
  const authError = useAppSelector((s) => s.auth.error);
  const dispatch = useAppDispatch();

  function handleSubmit(evt: any) {
    evt.preventDefault();
    dispatch(login(formData.username, formData.password));
  }

  let formErrors: any = {};
  if (authError) {
    formErrors = parseFormErrors(authError);
  }

  return (
    <Box sx={{ display: "grid", justifyItems: "center" }}>
      <Typography variant="h3" sx={{ mt: 2 }}>
        Login
      </Typography>
      {formErrors.message ? (
        <Alert severity="error">{formErrors.message}</Alert>
      ) : null}
      <Box
        component="form"
        autoComplete="off"
        sx={{
          display: "grid",
          my: 3,
          px: 1,
          width: { xs: "100%", lg: "60%" },
          justifyItems: "center",
        }}
        onSubmit={handleSubmit}>
        <TextField
          fullWidth
          id="username"
          label="Username"
          variant="outlined"
          sx={{ my: 1 }}
          value={formData.username}
          onChange={handleChange}
        />
        <TextField
          fullWidth
          id="password"
          label="Password"
          type="password"
          variant="outlined"
          sx={{ my: 1 }}
          value={formData.password}
          onChange={handleChange}
        />
        <Button sx={{ my: 2 }} variant="contained" type="submit">
          Submit
        </Button>
      </Box>
    </Box>
  );
}

export default Login;
