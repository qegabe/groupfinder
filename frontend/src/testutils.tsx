import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { RootState, setupStore } from "./store";
import { Provider } from "react-redux";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import { LocalizationProvider } from "@mui/x-date-pickers";
import { AdapterDayjs } from "@mui/x-date-pickers/AdapterDayjs";
import { PreloadedState } from "@reduxjs/toolkit";

interface CustomRenderOptions {
  currentRoute?: string;
  routePath?: string;
  preloadedState?: PreloadedState<RootState>;
}

function customRender(
  ui: ReactElement,
  cOptions: CustomRenderOptions = {},
  options?: Omit<RenderOptions, "wrapper">
) {
  const store = setupStore(cOptions.preloadedState);
  function AllTheProviders({ children }: { children: React.ReactNode }) {
    return (
      <Provider store={store}>
        <MemoryRouter
          initialEntries={
            cOptions.currentRoute ? [cOptions.currentRoute] : undefined
          }>
          <LocalizationProvider dateAdapter={AdapterDayjs}>
            {cOptions.routePath ? (
              <Routes>
                <Route path={cOptions.routePath} element={children} />
              </Routes>
            ) : (
              children
            )}
          </LocalizationProvider>
        </MemoryRouter>
      </Provider>
    );
  }
  return render(ui, { wrapper: AllTheProviders, ...options });
}

export * from "@testing-library/react";
export { customRender as render };
