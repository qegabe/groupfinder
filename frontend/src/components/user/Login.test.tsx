import React from "react";
import { render, waitFor, screen, fireEvent } from "../../testutils";
import "@testing-library/jest-dom";
import GroupFinderApi from "../../api";
import Login from "./Login";

const testToken =
  "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1c2VybmFtZSI6IlRlc3RVc2VyIn0.XrNY2ZbmtAke3MgrqVFd4xRN2wQElzeyTvGMBx98UGs";

const mockLogin = jest
  .spyOn(GroupFinderApi, "login")
  .mockImplementation(async () => {
    return testToken;
  });

const mockGetUser = jest
  .spyOn(GroupFinderApi, "getUser")
  .mockImplementation(async () => {});

afterEach(() => {
  jest.clearAllMocks();
});

it("renders", async () => {
  render(<Login />);
  expect(screen.getByText("Login")).toBeInTheDocument();
});

it("matches snapshot", async () => {
  const { asFragment } = render(<Login />);
  expect(asFragment()).toMatchSnapshot();
});

it("lets user login", async () => {
  render(<Login />);
  const usernameInput = screen.getByLabelText("Username");
  const passwordInput = screen.getByLabelText("Password");
  const submitButton = screen.getByText("Submit");
  fireEvent.change(usernameInput, { target: { value: "TestUser" } });
  fireEvent.change(passwordInput, { target: { value: "12345" } });
  fireEvent.click(submitButton);
  await waitFor(() => {
    expect(mockLogin).toBeCalledWith("TestUser", "12345");
  });
  expect(mockGetUser).toBeCalledWith("TestUser");
});
