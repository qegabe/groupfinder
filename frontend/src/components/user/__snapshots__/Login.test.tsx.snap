// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1tx56hj"
  >
    <h3
      class="MuiTypography-root MuiTypography-h3 css-vhs57-MuiTypography-root"
    >
      Login
    </h3>
    <form
      autocomplete="off"
      class="MuiBox-root css-u2f8ta"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="username"
          id="username-label"
        >
          Username
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="username"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Username
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1nin81k-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for="password"
          id="password-label"
        >
          Password
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id="password"
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Password
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-x2xvh3-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
  </div>
</DocumentFragment>
`;
